qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -44,77 +43,85 @@
+    (simulators or real chips). It has some asynchronous properties, and some
+    blocking behavior. This is a common use-case example:
+    .. highlight::
+
+        try:
+            job = IBMQJob( ... )
+            job.submit() # <--- It won't block.
+
+            while not job.done(): # <---- It will return false if the job
+                                  # hasn't finished yet
+                pass # <---- or maybe... serialize to disk? .. whatever the user wants
+
+            result = job.result() # <--- This could block, but as the job is already
+                                  # finished, it won't in this case.
+        except JobError as ex:
+            log("Something wrong happened!: {}".format(ex))
+
+    About the status in case of Error or Exception:
+    If there's an exception from the IBMQuantumExperience API once a job has been submitted,
+    we cannot set the status of the Job to ERROR, because it might happen that the Job is actually
+    queued or running, and let's say that the exception is caused because of a temporary
+    out of service situation of the server, once this situation is restored, we might want to ask
+    again about the status of our Job, that could be in state: queued, running or done.
+    if something wrong happened to the Job, it's server responsability to answer an error response
+    to a status query from Qiskit.
+    But, if the exception is thrown before having a valid Job ID, there's nothing we can do,
+    with this Job instance, as we cannot query the server for information using this Job instance.
+
     Attributes:
         _executor (futures.Executor): executor to handle asynchronous jobs
         _final_states (list(JobStatus)): terminal states of async jobs
     """
     _executor = futures.ThreadPoolExecutor()
 
-    def __init__(self, qobj, api, is_device):
+    def __init__(self, api, is_device, qobj=None, job_id=None, backend_name=None,
+                 creation_date=None):
         """IBMQJob init function.
+        We can instantiate Jobs from two sources: A QObj, and an already submitted job returned by
+        the API servers.
 
         Args:
-            qobj (Qobj): job description
             api (IBMQuantumExperience): IBM Q API
             is_device (bool): whether backend is a real device  # TODO: remove this after Qobj
+            qobj (Qobj): The Quantum Object. If not present, is because we are probably getting
+                         a Job from the API server.
+            job_id (String): The Job ID of an already submitted job.
+            backend_name(String): The name of the backend that run the job.
+            creation_date(String): When the Job was run.
         """
         super().__init__()
-        self._qobj = qobj
+        # TODO No need for this conversion, just use the new equivalent members above
+        if qobj is not None:
+            old_qobj = qobj_to_dict(qobj, version='0.0.1')
+            self._job_data = {
```

Comment:
Can we ensure that `self._job_data` is always present after instantiation, even if it is an empty dict of `None` or a similar measure? `self.submit()` accesses it directly assuming it is present, and even though it seems those lines are not really reachable if it is not present (thanks to the `if` at the beginning of the function), it will be great for maintenance to be 100% sure that all instances have the same attributes, and that they are defined in `__init__`.